app/javascript/components/upload/ClusteringForm.js

```diff
@@ -0,0 +1,97 @@
+import React, { useEffect } from 'react'
+
+import UploadSteps from './UploadSteps'
+import ClusteringFileForm from './ClusteringFileForm'
+
+const DEFAULT_NEW_CLUSTER_FILE = {
+  is_spatial: false,
+  file_type: 'Cluster'
+}
+
+/** Renders a form for uploading one or more cluster/spatial files */
+export default function ClusteringUploadForm({
+  formState,
+  addNewFile,
+  updateFile,
+  saveFile,
+  deleteFile,
+  handleSaveResponse
+}) {
+
+  const clusterFiles = formState.files.filter(UploadSteps.clustering.fileFilter)
+  const associatedClusterFileOptions = clusterFiles.filter(file => !file.is_spatial)
+    .map(file => ({ label: file.name, value: file._id }))
+
+  /** handle a change in the associated cluster select */
+  function updateCorrespondingClusters(file, val) {
+    let newVal = []
+    if (val) {
+      newVal = val.map(opt => opt.value)
+    }
+    updateFile(file._id, { spatial_cluster_associations: newVal })
+  }
+
+  useEffect(() => {
+    if (clusterFiles.length === 0) {
+      addNewFile(DEFAULT_NEW_CLUSTER_FILE)
+    }
+  }, [clusterFiles.length])
+
+  return <div>
```

Comment:
I like how this is encapsulated now - it seems like the form logic is getting a lot simpler with the componentization.